backend/creditsWorker.js

```diff
@@ -0,0 +1,190 @@
+      const balance = (previousBalance - creditsUsed).toString(10)
+
+      // Prepare the new history entry
+      granularHistory.unshift({
+        type: 'charge',
+        date,
+        sizeTotal: amount,
+        credits: creditsUsed.toString(10),
+        period: `${new Date(now - timeElapsed).toISOString()}/${date}`,
+        balance
+      })
+
+      // Persist the updated balance (optimization: faster reads elsewhere)
+      await sbp('chelonia.db/set', `_private_ownerCreditBalance_${billableEntity}`, balance)
+    } else if (type === 'credit') {
+      // 'amount' is the credit amount here
+      const creditsToAdd = BigInt(amount)
+      const newBalance = previousBalance + creditsToAdd
+      const newBalanceStr = newBalance.toString(10)
+
+      // Prepare the new history entry for credit addition.
+      const newEntry = {
+        type: 'credit',
+        date,
+        credits: creditsToAdd.toString(10), // Credits added
+        balance: newBalanceStr
+      }
+
+      granularHistory.unshift(newEntry)
+
+      // Persist the updated balance
+      await sbp('chelonia.db/set', `_private_ownerCreditBalance_${billableEntity}`, newBalanceStr)
+    } else {
+      console.error(`Invalid update type "${type}" for ${billableEntity}`)
+      return // Don't save if type is invalid
+    }
+
+    // Do we need to update the coarse history?
+    const lastCoarseSyncIdx = granularHistory.findIndex((value) => {
+      return !!(value.coarseSyncPoint)
+    })
+
+    if (lastCoarseSyncIdx >= GRANULAR_ENTRIES_PER_COARSE_ENTRIES || lastCoarseSyncIdx < 0) {
+      const coarseHistoryKey = `_private_ownerBalanceHistoryCoarse_${billableEntity}`
+      granularHistory[0].coarseSyncPoint = true
+      const coarseHistoryList = await sbp('chelonia.db/get', coarseHistoryKey, { bypassCache: true }) ?? '[]'
+      const coarseHistory = JSON.parse(coarseHistoryList)
+
+      if (coarseHistory.length >= COARSE_HISTORY_MAX_ENTRIES) {
+        coarseHistory.splice(COARSE_HISTORY_MAX_ENTRIES)
+      }
+
+      const { periodStart, charges, credits, periodSize, totalPeriodLength } = granularHistory.slice(0, lastCoarseSyncIdx < 0 ? granularHistory.length : lastCoarseSyncIdx).reduce((acc, entry) => {
+        if (entry.type === 'charge') {
+          acc.charges += BigInt(entry.credits)
+          const [periodStart, periodEnd] = entry.period.split('/')
+          const [periodStartDate, periodEndDate] = [Date.parse(periodStart), Date.parse(periodEnd)]
+          const periodLength = Math.floor(periodEndDate - periodStartDate)
+          // Avoid NaN propagation
+          if (periodLength >= 0) {
+            acc.periodSize += entry.sizeTotal * periodLength
+            acc.totalPeriodLength += periodLength
+          }
+          acc.periodStart = periodStart
+        } else if (entry.type === 'credit') {
+          acc.credits += BigInt(entry.credits)
+        } else {
+          throw new Error('Invalid entry type: ' + entry.type)
+        }
```

Comment:
Should we really throw here or just `console.error` log it?

If for some reason a small amount got corrupted this would permanently prevent the rest of the code from doing its job...

Reply:
Fair point. OTOH, not sure if it's wise to proceed in this case as the entries are... corrupted.

Reply:
Ah, ok, we can leave it as-is then. Please see my other comments tho about logging errors